tool-openssl/req.cc

```diff
@@ -0,0 +1,637 @@
+    return false;
+  }
+
+  // Set up X509V3 context for certificate
+  X509V3_CTX ctx;
+  X509V3_set_ctx_nodb(&ctx);
+  X509V3_set_ctx(&ctx, cert, cert, NULL, NULL, 0);  // Self-signed
+  X509V3_set_nconf(&ctx, conf.get());
+
+  // Add extensions from config to the certificate
+  bool result = X509V3_EXT_add_nconf(conf.get(), &ctx, "v3_ca", cert) != 0;
+
+  return result;
+}
+
+// Generate a random serial number for a certificate
+static bool generate_serial(X509 *cert) {
+  // Create a new BIGNUM to hold the random value
+  bssl::UniquePtr<BIGNUM> bn(BN_new());
+  if (!bn) {
+    fprintf(stderr, "Failed to create BIGNUM for serial\n");
+    return false;
+  }
+
+  // Generate random number with 159 bits (same as OpenSSL's SERIAL_RAND_BITS)
+  constexpr int SERIAL_RAND_BITS = 159;
+  if (!BN_rand(bn.get(), SERIAL_RAND_BITS, BN_RAND_TOP_ANY, BN_RAND_BOTTOM_ANY)) {
+    fprintf(stderr, "Failed to generate random serial number\n");
+    return false;
+  }
+
+  // Get a pointer to the certificate's ASN1_INTEGER serial number field
+  ASN1_INTEGER *serial = X509_get_serialNumber(cert);
+  if (!serial) {
+    fprintf(stderr, "Failed to get certificate serial number field\n");
+    return false;
+  }
+
+  // Convert BIGNUM to ASN1_INTEGER and set it as the certificate's serial number
+  if (!BN_to_ASN1_INTEGER(bn.get(), serial)) {
+    fprintf(stderr, "Failed to convert BIGNUM to ASN1_INTEGER\n");
+    return false;
+  }
+
+  return true;
+}
+
+bool reqTool(const args_list_t &args) {
+  args_map_t parsed_args;
+  args_list_t extra_args;
+  if (!ParseKeyValueArguments(parsed_args, extra_args, args, kArguments) ||
+      extra_args.size() > 0) {
+    PrintUsage(kArguments);
+    return false;
+  }
+
+  std::string newkey, subj, keyout, out;
+  unsigned int days;
+  bool help = false, new_flag = false, x509_flag = false, nodes = false;
+
+  GetBoolArgument(&help, "-help", parsed_args);
+  GetBoolArgument(&new_flag, "-new", parsed_args);
+  GetBoolArgument(&x509_flag, "-x509", parsed_args);
+  GetBoolArgument(&nodes, "-nodes", parsed_args);
+
+  GetString(&newkey, "-newkey", "", parsed_args);
+  GetUnsigned(&days, "-days", 30u, parsed_args);
+  GetString(&subj, "-subj", "", parsed_args);
+  GetString(&keyout, "-keyout", "" , parsed_args);
+  GetString(&out, "-out", "" , parsed_args);
+
+  if (help) {
+    PrintUsage(kArguments);
+    return false;
+  }
+
+  if (!new_flag && !x509_flag && newkey.empty()) {
+    fprintf(stderr, "Error: Missing required options, -x509, -new, or -newkey must be specified. \n");
+    return false;
+  }
+
+  std::string keyspec = "rsa:2048";
+  if (!newkey.empty()) {
+	  keyspec = newkey;
+  }
+
+  bssl::UniquePtr<EVP_PKEY> pkey(generate_key(keyspec.c_str()));
+  if (!pkey) {
+    fprintf(stderr, "Error: Failed to generate private key.\n");
+    return false;
+  }
+
+  // Generate and write private key
+  EVP_CIPHER *cipher = NULL;
+  if (!nodes) {
+	  cipher = (EVP_CIPHER*)EVP_des_ede3_cbc();
```

Comment:
Can we just directly use `const EVP_CIPHER` here to avoid the extra cast? `PEM_write_bio_PrivateKey` also seem to take a `const EVP_CIPHER` below.